core/kernel.go

```diff
@@ -73,300 +65,305 @@ const (
 	Logger         *logging.Logger
+	txCodec        txs.Codec
+	exeOptions     []execution.ExecutionOption
+	exeChecker     execution.BatchExecutor
+	exeCommitter   execution.BatchCommitter
+	keyClient      keys.KeyClient
+	keyStore       *keys.KeyStore
 	nodeInfo       string
 	processes      map[string]process.Process
 	shutdownNotify chan struct{}
 	shutdownOnce   sync.Once
 }
 
-func NewBurrowDB(dbDir string) dbm.DB {
+func newBurrowDB(dbDir string) dbm.DB {
 	return dbm.NewDB(BurrowDBName, dbm.GoLevelDBBackend, dbDir)
 }
 
-func NewKernel(ctx context.Context, keyClient keys.KeyClient, privValidator tmTypes.PrivValidator,
-	genesisDoc *genesis.GenesisDoc, tmConf *tmConfig.Config, rpcConfig *rpc.RPCConfig, keyConfig *keys.KeysConfig,
-	keyStore *keys.KeyStore, exeOptions []execution.ExecutionOption, authorizedPeersProvider abci.PeersFilterProvider, restore string, nodeKey *crypto.PrivateKey, logger *logging.Logger) (*Kernel, error) {
-
-	var err error
-	kern := &Kernel{
+// NewKernel initializes an empty kernel
+func NewKernel() (*Kernel, error) {
+	runID, err := simpleuuid.NewTime(time.Now()) // Create a random ID based on start time
+	if err != nil {
+		return nil, err
+	}
+	return &Kernel{
+		RunID:          runID,
 		processes:      make(map[string]process.Process),
 		shutdownNotify: make(chan struct{}),
-	}
-	// Create a random ID based on start time
-	kern.RunID, err = simpleuuid.NewTime(time.Now())
+		txCodec:        txs.NewAminoCodec(),
+	}, nil
+}
 
-	logger = logger.WithScope("NewKernel()").With(structure.TimeKey, log.DefaultTimestampUTC,
-		structure.RunId, kern.RunID.String())
-	tendermintLogger := logger.With(structure.CallerKey, log.Caller(LoggingCallerDepth+1))
+// SetLogger initializes the kernel with the provided logger
+func (kern *Kernel) SetLogger(logger *logging.Logger) {
+	logger = logger.WithScope("NewKernel()").With(structure.TimeKey,
+		log.DefaultTimestampUTC, structure.RunId, kern.RunID.String())
 	kern.Logger = logger.WithInfo(structure.CallerKey, log.Caller(LoggingCallerDepth))
-	stateDB := NewBurrowDB(tmConf.DBDir())
+	kern.Emitter = event.NewEmitter(kern.Logger)
+}
 
+func (kern *Kernel) LoadBlockchain(genesisDoc *genesis.GenesisDoc, dbDir string) (stateDB dbm.DB, err error) {
+	stateDB = newBurrowDB(dbDir)
 	kern.Blockchain, err = bcm.LoadOrNewBlockchain(stateDB, genesisDoc, kern.Logger)
 	if err != nil {
 		return nil, fmt.Errorf("error creating or loading blockchain state: %v", err)
 	}
-
+	kern.Blockchain.SetBlockStore(bcm.NewBlockStore(blockchain.NewBlockStore(stateDB)))
 	heightValuer := log.Valuer(func() interface{} { return kern.Blockchain.LastBlockHeight() })
 	kern.Logger = kern.Logger.With("height", heightValuer)
-	tendermintLogger = tendermintLogger.With("height", heightValuer)
+	return stateDB, nil
+}
+
+// LoadState starts from scratch or previous chain
+func (kern *Kernel) LoadState(genesisDoc *genesis.GenesisDoc, dbDir string) (err error) {
+	stateDB, err := kern.LoadBlockchain(genesisDoc, dbDir)
```

Comment:
The way this is now is technically fine, but it looks like we could end up trying to set up `DB` instance on the same directory twice. To avoid it seeming this way, or being refactored so it does in the future only one method should be responsible for setting up the DB. So we may want to pass the `dbm.DB` instance into both `LoadState` and `LoadBlockchain`

Actually no, we should set up the DB in `NewKernel`